### PR TITLE
stats: adjust datum type when using the index query feedback

### PR DESCRIPTION
**Pull request by lamxTyler**
### What problem does this PR solve? 

It is discovered in https://github.com/pingcap/tidb/pull/10606. The ranges in the feedback are the result of `EncodeKey` of origin datums, however, different types may be converted to the same type in encode, for example, `float32` will be encoded in `float64`. So the decoded scan range may not be the same type with the column, which may cause problems if we try to locate the datum in the column histogram, which assumes that the type of datum is the same with histograms.

### What is changed and how it works?

After decoding the ranges, it will convert the datum to column type.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - None

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: statistics/feedback.go [hunk 1]
```diff
@@ -811,24 +811,48 @@ func decodeFeedbackForPK(q *QueryFeedback, pb *queryFeedback, isUnsigned bool) {
 	}
 }
 
-func decodeFeedbackForColumn(q *QueryFeedback, pb *queryFeedback) error {
+// ConvertDatumsType converts the datums type to `ft`.
+func ConvertDatumsType(vals []types.Datum, ft *types.FieldType, loc *time.Location) error {
```
Comment (winoros) [inline statistics/feedback.go]: Need we export it?

Comment (lamxTyler) [reply statistics/feedback.go]: Yes, it is also used in `convertRangeType`, which is in package `handle`.

File: statistics/handle/update.go [hunk 2]
```diff
@@ -1042,11 +1053,14 @@ func (h *Handle) DumpFeedbackForIndex(q *statistics.QueryFeedback, t *statistics
 		rangeFB := &statistics.QueryFeedback{PhysicalID: q.PhysicalID}
 		// prefer index stats over column stats
 		if idx := t.IndexStartWithColumn(colName); idx != nil && idx.Histogram.Len() != 0 {
-			rangeCount, err = t.GetRowCountByIndexRanges(sc, idx.ID, []*ranger.Range{&rang})
+			rangeCount, err = t.GetRowCountByIndexRanges(sc, idx.ID, []*ranger.Range{rang})
 			rangeFB.Tp, rangeFB.Hist = statistics.IndexType, &idx.Histogram
 		} else if col := t.ColumnByName(colName); col != nil && col.Histogram.Len() != 0 {
-			rangeCount, err = t.GetRowCountByColumnRanges(sc, col.ID, []*ranger.Range{&rang})
-			rangeFB.Tp, rangeFB.Hist = statistics.ColType, &col.Histogram
+			err = convertRangeType(rang, col.Tp, sc.TimeZone)
```
Comment (zz-jason) [inline statistics/handle/update.go]: should we use `time.UTC`?

Comment (lamxTyler) [reply statistics/handle/update.go]: Yes, in fact the `sc` defined in line 1014 is already `time.UTC`.

Comment (zz-jason) [reply statistics/handle/update.go]: I think directly pass `time.UTC` can improve the code readability.

**Pull request merged**